subsystems/mvdm/ntvdm/bios/vidbios.c

```diff
@@ -2942,6 +2942,295 @@ static VOID VidBiosPrintCharacter(CHAR Character, BYTE Attribute, BOOLEAN UseAtt
+            {
+                MaxColor = 1;
+                ValueShift = 7 - X % 8;
+                PixelsPerByte = 8;
+            }
+            else
+            {
+                MaxColor = 3;
+                ValueShift = (3 - X % 4) * 2;
+                PixelsPerByte = 4;
+            }
+
+            Xor = (Value & 0x80) != 0;
+            Value &= MaxColor;
+            Address = TO_LINEAR((Y & 1) ? CGA_ODD_VIDEO_SEG : CGA_EVEN_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + ((Y / 2) * ScreenWidth + X) / PixelsPerByte);
+
+            EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            if (Xor)
+            {
+                MemoryValue ^= Value << ValueShift;
+            }
+            else
+            {
+                MemoryValue &= ~(MaxColor << ValueShift);
+                MemoryValue |= Value << ValueShift;
+            }
+            EmulatorWriteMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            break;
+        }
+
+        case 0x0D: /* EGA 320x200 16c */
+        case 0x0E: /* EGA 640x200 16c */
+        case 0x10: /* EGA 640x350 16c */
+        case 0x12: /* VGA 640x480 16c */
+        {
+            ULONG Address;
+            BOOL Xor;
+            BYTE OldMode;
+            BYTE OldRotation;
+            BYTE OldRead;
+            BYTE OldMask;
+            BYTE CurrentPlaneMask;
+            BYTE CurrentPlane;
+            BYTE Shift = 7 - X % 8;
+
+            Xor = (Value & 0x80) != 0;
+            Value &= 15;
+            Address = TO_LINEAR(GRAPHICS_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + (Y * ScreenWidth + X) / 8);
+
+            /* Make sure we're in write mode 0 */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_MODE_REG);
+            OldMode = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, 0x00);
+
+            /* Setup the rotation */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_ROTATE_REG);
+            OldRotation = IOReadB(VGA_GC_DATA);
+            IOWriteB(VGA_GC_DATA, Xor ? 0x18 : 0x00);
+
+            /* Save old read */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+            OldRead = IOReadB(VGA_GC_DATA);
+
+            /* Save old mask */
+            IOWriteB(VGA_SEQ_INDEX, VGA_SEQ_MASK_REG);
+            OldMask = IOReadB(VGA_SEQ_DATA);
+
+            /* Loop and write over all four color bitplanes */
+            CurrentPlaneMask = 1;
+            for (CurrentPlane = 0; CurrentPlane < 4; CurrentPlane++)
+            {
+                BYTE Dummy;
+                BYTE ActualValue;
+
+                IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+                IOWriteB(VGA_GC_DATA, CurrentPlane);
+                IOWriteB(VGA_SEQ_DATA, CurrentPlaneMask);
+                EmulatorReadMemory(&EmulatorContext, Address, &Dummy, sizeof(BYTE));
+
+                /* There's probably a way to handle this logic on write, but I'll just leave it at that */
```

Comment:
I wonder whether any of these two patches in the JIRA report implement that kind of logic, or whether what is done for these modes in VidBiosDrawGlyph() is what could be done.